### DISCUSSION TITLE
C# files are incorrectly detected as Smalltalk

### DISCUSSION THREAD

**Discussion by daviddesmet** [github-linguist/linguist#5869] [Classification] [answered]
Some C# files are incorrectly detected as Smalltalk.

## Describe the bug
Some C# files are incorrectly detected as Smalltalk as seen here https://github.com/daviddesmet/paseto-dotnet/search?l=smalltalk

<img width="564" alt="image" src="https://user-images.githubusercontent.com/2607383/163864825-5e28259b-8d46-490d-acf5-511c62ea9cf3.png">

<img width="495" alt="image" src="https://user-images.githubusercontent.com/2607383/163864177-356af1d0-d6f8-45ac-a443-ff8ebbda2da3.png">

### Expected behaviour
It should be detected as C#

**Comment by lildude** [comment]
Smalltalk and C# share the `.cs` extension so these files should be caught by the heuristic:

https://github.com/github/linguist/blob/4ffcdbcbb60a74cbfbd37656bcc3fcea4eca8e26/lib/linguist/heuristics.yml#L127-L132

If they're not, which appears to be the case here, things fall through to the classifier which makes a guess based on the samples we have for each langauge and compares it to the tokens found in yur files.

I've just cloned your repo and run linguist against it and only one of those files is classified as Smalltalk:

```console
$ github-linguist . --breakdown | tail -5

Smalltalk:
src/Paseto/Cryptography/Internal/Blake2bNormal.cs
tests/Paseto.Tests/Crypto/Ed25519TestVectors.cs

$
```

Both of these are falling through to the classifier and it's getting it wrong.

This last file is massive (1.4M) which explains the high percentage of Smalltalk.

The difference between what you see on GitHub.com and what I'm seeing locally will be down to the Search index which is completely independent from Linguist. 

Your overrides aren't having any effect either as the paths of the files has changed.

What I'm not sure about is why these files are falling through the heuristic 🤔

**Reply by lildude** [reply, marked as answer]
Oooo, just spotted it, and it affects quite a few of your files, though they're getting correctly identified by the classifier as they look like the other C# files we've got samples for.

The issue is these files are using file scoped namespaces which were introduced in C# 10 but the heurstic which pre-dates C# 10 by about 6 years doesn't cater for this.

For example,

```C#
namespace Paseto.Cryptography;
```

... won't be matched by `^(\s*namespace\s*[\w\.]+\s*{|\s*\/\/)` because it doesn't have a `{`.

I'll open a new issue for this.

**Reply by lildude** [reply]
https://github.com/github/linguist/issues/5870

**Reply by daviddesmet** [reply]
Thanks a lot for digging into it and opening an issue. :)
